### PR TITLE
[SC] Separate CallData deserialization

### PR DESCRIPTION
**Pull request by rowandh**
Separates the deserialization of CallData from SmartContractCarrier.

### CODE REVIEW

File: src/Stratis.SmartContracts.Executor.Reflection/CallData.cs [hunk 1]
```diff
@@ -1,38 +1,44 @@
...
+using Stratis.SmartContracts.Core;
 
 namespace Stratis.SmartContracts.Executor.Reflection
 {
     public struct CallData
     {
-        public CallData(byte opCodeType, int vmVersion, ulong gasPrice, Gas gasLimit, uint160 address,
-            string method, string parameters = "")
+        public CallData(int vmVersion, ulong gasPrice, Gas gasLimit, uint160 address,
+            string method, string rawParameters = "", object[] methodParameters = null)
```
[the remark this answers is not in the document]
Comment (rowandh) [reply src/Stratis.SmartContracts.Executor.Reflection/CallData.cs]: You ask for the impossible :) string.Empty is not a compile-time constant

File: src/Stratis.Bitcoin.Features.SmartContracts/ReflectionExecutor/Consensus/Rules/SmartContractFormatRule.cs [hunk 2]
```diff
@@ -57,36 +57,50 @@ private void CheckTransaction(Transaction transaction, Money suppliedBudget)
...
+
+            var callDataDeserializationResult = serializer.Deserialize(scTxOut.ScriptPubKey.ToBytes());
+
+            if (callDataDeserializationResult.IsFailure)
+            {
+                new ConsensusError("invalid-calldata-format", "Invalid CallData format").Throw();
+            }
+
+            var callData = callDataDeserializationResult.Value;
+
```
[the remark this answers is not in the document]
Comment (rowandh) [reply src/Stratis.Bitcoin.Features.SmartContracts/ReflectionExecutor/Consensus/Rules/SmartContractFormatRule.cs]: I'm confused by this one. Our own guidelines say `Always provide a blank line after a curly brace that is not followed by another curly brace.` And the last blank line before the if cause is consistent with the rest of the code.

File: src/Stratis.SmartContracts.Executor.Reflection/CallData.cs [hunk 3]
```diff
@@ -1,38 +1,44 @@
...
+            this.MethodParameters = methodParameters;
             this.ContractExecutionCode = new byte[0];
         }
 
-        public CallData(byte opCodeType, int vmVersion, ulong gasPrice, Gas gasLimit, byte[] code, string parameters = "")
+        public CallData(int vmVersion, ulong gasPrice, Gas gasLimit, byte[] code, string rawParameters = "",
+            object[] methodParameters = null)
         {
-            this.OpCodeType = opCodeType;
+            this.OpCodeType = (byte)ScOpcodeType.OP_CREATECONTRACT;
```
[the remark this answers is not in the document]
Comment (fassadlr) [reply src/Stratis.SmartContracts.Executor.Reflection/CallData.cs]: 👍

File: src/Stratis.Bitcoin.Features.SmartContracts/ReflectionExecutor/Consensus/Rules/SmartContractFormatRule.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.SmartContracts/ReflectionExecutor/Consensus/Rules/SmartContractFormatRule.cs]: Oh sorry yeah, its actually more of personal preference but you are right we should follow guidelines.

File: src/Stratis.SmartContracts.Executor.Reflection/CallDataSerializer.cs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            var byteCursor = 1;
+            var takeLength = 0;
+
+            var type = smartContractBytes[0];
+
+            var vmVersion = Deserialize<int>(smartContractBytes, ref byteCursor, ref takeLength);
+            var gasPrice = (Gas)Deserialize<ulong>(smartContractBytes, ref byteCursor, ref takeLength);
+            var gasLimit = (Gas)Deserialize<ulong>(smartContractBytes, ref byteCursor, ref takeLength);
+
+            if (type == (byte)ScOpcodeType.OP_CALLCONTRACT)
```
[the remark this answers is not in the document]
Comment (fassadlr) [reply src/Stratis.SmartContracts.Executor.Reflection/CallDataSerializer.cs]: I was actually referring to all the logic contained in the `if` ;)

File: src/Stratis.SmartContracts.Executor.Reflection/CallData.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (fassadlr) [reply src/Stratis.SmartContracts.Executor.Reflection/CallData.cs]: Doh. My bad.

File: src/Stratis.SmartContracts.Executor.Reflection/CallDataSerializer.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (rowandh) [reply src/Stratis.SmartContracts.Executor.Reflection/CallDataSerializer.cs]: That doesn't make much sense to me as it makes it harder to follow what's going on. Deserialization happens in order and this reflects that.

**Comment by rowandh** [comment]
@fassadlr I've run all smart contracts unit tests + smart contracts integration tests successfully

**Pull request merged**